src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,763 @@
+#include "expression_evaluator.hpp"
+
+#include <iterator>
+#include <type_traits>
+
+#include "boost/variant/apply_visitor.hpp"
+
+#include "expression/abstract_expression.hpp"
+#include "expression/array_expression.hpp"
+#include "expression/abstract_predicate_expression.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/case_expression.hpp"
+#include "expression/expression_factory.hpp"
+#include "all_parameter_variant.hpp"
+#include "expression/arithmetic_expression.hpp"
+#include "expression/exists_expression.hpp"
+#include "expression/extract_expression.hpp"
+#include "expression/function_expression.hpp"
+#include "expression/logical_expression.hpp"
+#include "expression/in_expression.hpp"
+#include "expression/pqp_column_expression.hpp"
+#include "expression/pqp_select_expression.hpp"
+#include "expression/value_expression.hpp"
+#include "operators/abstract_operator.hpp"
+#include "storage/materialize.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "sql/sql_query_plan.hpp"
+#include "resolve_type.hpp"
+#include "utils/lambda_visitor.hpp"
+#include "storage/value_column.hpp"
+#include "utils/assert.hpp"
+#include "expression_functors.hpp"
+
+using namespace std::string_literals;
+using namespace opossum::expression_factory;
+
+namespace opossum {
+
+ExpressionEvaluator::ExpressionEvaluator(const std::shared_ptr<const Chunk>& chunk):
+_chunk(chunk)
+{
+  _output_row_count = _chunk->size();
+  _column_materializations.resize(_chunk->column_count());
+}
+
+template<typename T>
+ExpressionResult<T> ExpressionEvaluator::evaluate_expression(const AbstractExpression& expression) {
+  switch (expression.type) {
+    case ExpressionType::Arithmetic:
+      return evaluate_arithmetic_expression<T>(static_cast<const ArithmeticExpression&>(expression));
+
+    case ExpressionType::Logical:
+      return evaluate_logical_expression<T>(static_cast<const LogicalExpression&>(expression));
+
+    case ExpressionType::Predicate: {
+      const auto& predicate_expression = static_cast<const AbstractPredicateExpression&>(expression);
+
+      if (is_lexicographical_predicate_condition(predicate_expression.predicate_condition)) {
+        return evaluate_binary_predicate_expression<T>(static_cast<const BinaryPredicateExpression&>(expression));
```

Comment:
As much as I am in favor of speaking method names, I think we can do with one "expression" less here. `evaluate_binary_predicate` should be just fine.